### DISCUSSION TITLE
Thoughts on Unit-Testing

### DISCUSSION THREAD

**Discussion by KDGundermann** [twinbasic/twinbasic#663] [General]
I started to migrate the first tests from my application where I used VBUnit / AccUnit .

Some remarks:
1. Will there be any Setup/Teardown (or TestInitialize/TestCleanup) Attributes for methods ?
   (and maybe a TestIgnore )

2. For the folks who is used to work with nunit [TestFixture] is known attribute
but coming from JUnit, pytest or Ruby : 
  > A test fixture is a fixed state of a set of objects used as a baseline for running tests

  See also: [junit4](https://github.com/junit-team/junit4/wiki/test-fixtures) [RubyOnRails](https://api.rubyonrails.org/v3.1/classes/ActiveRecord/Fixtures.html)
  so I would prefer to have [TestClass] or [TestModule] as attribute

3. Using Assert everytime with a Strict/Exact/Permissive is little bit to verbose IMHO
( and what is the difference between `Assert.Exact.IsNotNothing obj`  and `Assert.Strict.IsNotNothing obj`  ?)
I would prefer to have an optional param to the Assert Methods:
e.g `Assert.IsEqual 5, myValue, "Wrong Value", vbCompareExact`

4. I'm not a fan of the reversed params in AssertEqual
  Usually I write `If myValue = 5 Then .. ` but I have to write `Assert.IsEqual 5, myValue`

just my thoughts from today

**Comment by WaynePhillipsEA** [comment]
Thanks for checking out the unit testing support!

1. Yes, these are planned

For the other points, I would like to hear from others here for their opinions.  I know some others here are using the unit testing in tB, so hopefully they will have a look at your comments and give some feedback.  

If there's a general agreement for change, I'll certainly look at it since nothing is set in stone at this stage.

**Comment by retailcoder** [comment]
> Usually I write If myValue = 5 Then .. but I have to write Assert.IsEqual 5, myValue

I only used a few frameworks and all were .NET so maybe that's a bias right there, but I expect to supply expected and actual values in that exact order. Rubberduck unit tests work that way too, since MSTest was the inspiration.

**Reply by WaynePhillipsEA** [reply]
As is probably obvious, I mostly copied Rubberduck when creating the tB version 😁

**Comment by KDGundermann** [comment]
with [SimplyVBUnit](https://github.com/kellyethridge/SimplyVBUnit) there is an open source, simple to use test framework for VB6 with 20k+ downloads.

Examples for tests:
```
    ' Numeric  
    Assert.That Actual, Iz.LessThan(Expected), "This should be lesser." 
    ' Double
    Assert.That Actual, Iz.EqualTo(Expected).Within(0.0001)
    Assert.AreEqualFloats Expected, Actual, 0.0001
    ' Collection
    Assert.That Actual, Has.Member("b") 
    ' String
    Assert.That Actual, Iz.StringStarting(Expected).IgnoreCase
```

**Reply by retailcoder** [reply]
I'd fight the inclusion of "iz" in tB with everything I've got, tbh. Kelly's work is impressive, but I've always found the "Assert.That" syntax rather annoyingly awkward and unintuitive. I'm surprised a repo with so few stars, forks, and contributors can boast 20K downloads, but if I let a Rubberduck release accumulate downloads for 3 years it'd probably be around 20K too, given we consistently hit 10K per ~6 months.

**Reply by retailcoder** [reply]
https://somsubhra.github.io/github-release-stats/?username=KellyEthridge&repository=SimplyVBUnit&page=1&per_page=5

That's 200, not 20K.

**Comment by KDGundermann** [comment]
Is there a better way to write test for <greater than> or <less than> ?
![image](https://user-images.githubusercontent.com/208315/150828009-8159870e-7d6d-45bc-8836-da2240aa0c9b.png)
